### PR TITLE
🔒 Generic Oauth installer

### PR DESCRIPTION
**Pull request by awtkns**
- Adds a generic interface for external apps to authenticate and integrate with our system
- Adds a slack integration as first demo

We should probably encrypt the access tokens in the database, that can be future

### CODE REVIEW

File: next/prisma/schema.prisma [hunk 1]
```diff
@@ -99,6 +99,39 @@ model VerificationToken {
...
+  update_date DateTime? @updatedAt
+  delete_date DateTime?
+
+  @@unique([installation_id])
+  @@map("oauth_credentials")
+}
+
+model OAuthInstallation {
+  id              String  @id @default(cuid())
+  user_id         String
```
Comment (shahrishabh7) [inline next/prisma/schema.prisma:121]: do we need to be more specific with id? is it an account identifier?

**Review by shahrishabh7 — APPROVED** [review]

**Comment by shahrishabh7** [comment]
pytest is failing - `ERROR  - ModuleNotFoundError: No module named 'slack_sdk'`

File: next/prisma/schema.prisma (hunk 1 above)
Comment (awtkns) [reply next/prisma/schema.prisma:121]: nope, it maps directly to the user table :)

File: next/prisma/schema.prisma [hunk 2]
```diff
@@ -99,6 +99,39 @@ model VerificationToken {
   @@unique([identifier, token])
 }
 
+model OAuthCredentials {
+  id              String  @id @default(cuid())
+  installation_id String
+  provider        String
+  token_type      String
+  access_token    String
+  scope           String?
```
Comment (asim-shrestha) [inline next/prisma/schema.prisma:108]: How does this typically work

File: platform/reworkd_platform/web/api/auth/views.py [hunk 3]
```diff
@@ -44,3 +48,24 @@ async def pusher_authentication(
...
+
+
+@router.get("/{provider}")
+async def oauth_install(
+    user: UserBase = Depends(get_current_user),
+    installer: OAuthInstaller = Depends(installer_factory),
+) -> str:
+    """Install an OAuth App"""
+    url = await installer.install(user)
+    print(url)
```
Comment (asim-shrestha) [inline platform/reworkd_platform/web/api/auth/views.py:60]: nit: print

File: platform/reworkd_platform/tests/test_oauth_installers.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+def test_installer_factory(mocker):
+    crud = mocker.Mock()
+    installer_factory("slack", crud)
+
+
+def test_integration_dne(mocker):
+    crud = mocker.Mock()
+
+    with pytest.raises(NotImplementedError):
+        installer_factory("asim", crud)
```
Comment (asim-shrestha) [inline platform/reworkd_platform/tests/test_oauth_installers.py:15]: 🤨

File: platform/reworkd_platform/settings.py [hunk 5]
```diff
@@ -102,6 +102,11 @@ class Settings(BaseSettings):
     ff_mock_mode_enabled: bool = False  # Controls whether calls are mocked
     max_loops: int = 25  # Maximum number of loops to run
 
+    # Settings for slack
+    slack_client_id: str = ""
+    slack_client_secret: str = ""
+    slack_redirect_uri: str = ""
+
```
Comment (asim-shrestha) [inline platform/reworkd_platform/settings.py:109]: Shouldn't this be in the DB now?

File: platform/reworkd_platform/db/crud/base.py [hunk 6]
```diff
@@ -1,5 +1,9 @@
+from typing import TypeVar
+
 from sqlalchemy.ext.asyncio import AsyncSession
 
+T = TypeVar("T", bound="BaseCrud")
```
Comment (asim-shrestha) [inline platform/reworkd_platform/db/crud/base.py:5]: What does this do?

File: platform/reworkd_platform/settings.py (hunk 5 above)
Comment (awtkns) [reply platform/reworkd_platform/settings.py:109]: No this is for our app, we are installing our app into other peoples slack

**Pull request merged**